src/drivers/camera_capture/camera_capture.cpp

```diff
@@ -0,0 +1,313 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file camera_capture.cpp
+ *
+ * Online and offline geotagging from camera feedback
+ *
+ * @author Mohammed Kabir <kabir@uasys.io>
+ */
+
+#include "camera_capture.hpp"
+
+#define commandParamToInt(n) static_cast<int>(n >= 0 ? n + 0.5f : n - 0.5f)
+
+namespace camera_capture
+{
+CameraCapture	*g_camera_capture;
+}
+
+CameraCapture::CameraCapture() :
+	_capture_enabled(false),
+	_trigger_pub(nullptr),
+	_command_ack_pub(nullptr),
+	_command_sub(-1),
+	_camera_capture_feedback(false),
+	_camera_capture_mode(0),
+	_camera_capture_edge(0),
+	_capture_seq(0),
+	_last_fall_time(0),
+	_last_exposure_time(0),
+	_capture_overflows(0)
+{
+
+	memset(&_work, 0, sizeof(_work));
+
+	// Parameters
+	_p_strobe_delay = param_find("CAM_CAP_DELAY");
+	param_get(_p_strobe_delay, &_strobe_delay);
+
+	_p_camera_capture_feedback = param_find("CAM_CAP_FBACK");
+	param_get(_p_camera_capture_feedback, &_camera_capture_feedback);
+
+	_p_camera_capture_mode = param_find("CAM_CAP_MODE");
+	param_get(_p_camera_capture_mode, &_camera_capture_mode);
+
+	_p_camera_capture_edge = param_find("CAM_CAP_EDGE");
+	param_get(_p_camera_capture_edge, &_camera_capture_edge);
+
+	if (_camera_capture_feedback) {
+		struct camera_trigger_s trigger = {};
+		_trigger_pub = orb_advertise(ORB_ID(camera_trigger), &trigger);
+	}
+}
+
+CameraCapture::~CameraCapture()
+{
+	camera_capture::g_camera_capture = nullptr;
+}
+
+void
+CameraCapture::capture_callback(uint32_t chan_index,
```

Comment:
Assuming this is the ISR call back. Long service is not recommended on a timer call back. So off load the values to a queue or buffer. Then scheduled HP work to read the buffer and do the logic and publication.

Reply:
The also facilitates the lazy advertise as it can be done in the context of a work thread and not an ISR

Reply:
I pushed a fix for this.. Now I just push to a buffer in the ISR callback and then schedule the rest in the work queue. On average it is 2 or 3 micro seconds faster than before.